File: examples/showcase.cs
/**
 * Hatsune Miku Theme - C# Showcase
 * All-Miku Synthesis: Every voice, one stage.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Text.RegularExpressions;

namespace MikuTheme.Showcase;

// Interface: #B2EBE7 Bold (Ice Teal - Append Light)
public interface ISynthesizer
{
    // Method signature: #00BCD4 (NT Cyan)
    Task<string> SingAsync(string song);
    MikuVersion Version { get; }
}

// Generic interface
public interface IMetadataStore<T> where T : class
{
    void Set(string key, T value);
    T? Get(string key);
}

// Enum: #B2EBE7 Bold, EnumMember: #E05096
public enum MikuVersion
{
    V2Classic = 0,
    V3 = 1,
    V4X = 2,
    NT = 3,
    Sekai = 4,
    V6AI = 5
}

[Flags]
public enum StageMode
{
    None = 0,
    Idle = 1 << 0,
    Performing = 1 << 1,
    Holographic = 1 << 2
}

// Record (C# 9+): #B2EBE7 Bold
public record VoiceBank(
    string Name,
    MikuVersion Version,
    (int Min, int Max) FrequencyRange
)
{
    // Computed property
    public bool IsActive { get; init; } = true;
}

// Record struct (C# 10+)
public readonly record struct FrequencyRange(int Min, int Max);

// Attribute: #FFD740 Italic (Meta Amber)
[AttributeUsage(AttributeTargets.Method | AttributeTargets.Class)]
public class VocaloidAttribute : Attribute
{
    public string? EngineName { get; init; }
    public int ReleaseYear { get; init; }
}

// Custom exception: #FF5370
public class PerformanceException : Exception
{
    public PerformanceException(string message) : base(message) { }
    public PerformanceException(string message, Exception inner) : base(message, inner) { }
}

// Abstract class
public abstract class StageElement
{
    public abstract void Activate();

    // Virtual method with default implementation
    public virtual void Deactivate()
    {
        Console.WriteLine("Element deactivated");
    }
}

// Class: #B2EBE7 Bold (Ice Teal)
[Vocaloid(EngineName = "VOCALOID2", ReleaseYear = 2007)]
public class DigitalDiva : StageElement, ISynthesizer, IDisposable
{
    // Constants: #E05096 (Magenta LED - Rhythm)
    public const string CanonicalColor = "#39C5BB";
    private const int DefaultBpm = 39;
    private const int MaxEnergy = 100;

    // Static readonly field
    public static readonly IReadOnlyList<string> VoiceTypes =
        new[] { "dark", "soft", "light", "sweet", "vivid", "solid" };

    // Properties: #90B8B2
    public string Name { get; }
    public VoiceBank VoiceBank { get; }
    public MikuVersion Version => VoiceBank.Version;

    // Auto-property with init accessor
    public StageMode Mode { get; private set; } = StageMode.Idle;

    // Backing field pattern
    private int _energy = MaxEnergy;
    public int Energy
    {
        get => _energy;
        set => _energy = Math.Clamp(value, 0, MaxEnergy);
    }

    // Event: #90B8B2
    public event EventHandler<string>? SongStarted;
    public event EventHandler<string>? SongCompleted;

    // Constructor
    public DigitalDiva(string name = "Hatsune Miku", MikuVersion version = MikuVersion.V2Classic)
    {
        Name = name;
        VoiceBank = new VoiceBank(name, version, (80, 1100));
    }

    // Primary constructor (alternative style shown via record above)

    // Override method: #00BCD4
    public override void Activate()
    {
        Mode = StageMode.Performing;
        Console.WriteLine($"[MIKU] {Name} activated!");
    }

    // Async method: #00BCD4 (NT Cyan)
    public async Task<string> SingAsync(string song)
    {
        // Keywords: #39C5BB Bold
        if (Energy < 10)
        {
            throw new PerformanceException("Low energy: Please recharge with leeks");
        }

        // Raise event
        SongStarted?.Invoke(this, song);

        // String interpolation: #9CCC65 with #FF80AB expressions
        var message = $"[MIKU] Now singing: {song}";
        Console.WriteLine(message);

        // Await: Keywords #39C5BB Bold
        await Task.Delay(100);

        Energy -= 10;
        SongCompleted?.Invoke(this, song);

        return message;
    }

    // Expression-bodied member: #00BCD4
    public static string GetCanonicalColor() => CanonicalColor;

    // Generic method
    public T? GetMetadata<T>(string key) where T : class => default;

    // Operator overloading
    public static bool operator ==(DigitalDiva? left, DigitalDiva? right) =>
        left?.Name == right?.Name && left?.Version == right?.Version;

    public static bool operator !=(DigitalDiva? left, DigitalDiva? right) =>
        !(left == right);

    // Indexer
    public string this[int index] => VoiceTypes[index];

    // Deconstruct for pattern matching
    public void Deconstruct(out string name, out MikuVersion version)
    {
        name = Name;
        version = Version;
    }

    // IDisposable implementation
    private bool _disposed;

    public void Dispose()
    {
        if (!_disposed)
        {
            // Cleanup resources
            _disposed = true;
            GC.SuppressFinalize(this);
        }
    }

    // Override Equals and GetHashCode
    public override bool Equals(object? obj) =>
        obj is DigitalDiva other && this == other;

    public override int GetHashCode() =>
        HashCode.Combine(Name, Version);
}

// Extension methods
public static class DigitalDivaExtensions
{
    public static async Task PerformConcertAsync(
        this DigitalDiva diva,
        IEnumerable<string> songs)
    {
        foreach (var song in songs)
        {
            await diva.SingAsync(song);
        }
    }
}

// Static class
public static class MikuUtils
{
    // RegEx: #B388FF (Hologram Purple)
    private static readonly Regex MikuPattern =
        new(@"(?<name>Miku)[-_]?(?<version>V\d+)?", RegexOptions.Compiled);

    public static bool IsMikuRelated(string text) =>
        MikuPattern.IsMatch(text);
}

// Main program class
public class Program
{
    public static async Task Main(string[] args)
    {
        // Using declaration
        using var miku = new DigitalDiva("Hatsune Miku", MikuVersion.V2Classic);

        // Subscribe to events with lambda
        miku.SongStarted += (sender, song) =>
            Console.WriteLine($"Starting: {song}");

        // Collection expression (C# 12)
        List<string> songs = ["Melt", "World is Mine", "Rolling Girl"];

        // LINQ query syntax
        var longSongs = from song in songs
                        where song.Length > 4
                        orderby song
                        select song.ToUpper();

        // LINQ method syntax
        var shortSongs = songs
            .Where(s => s.Length <= 4)
            .Select(s => $"[{s}]")
            .ToList();

        // Pattern matching with switch expression
        var description = miku.Version switch
        {
            MikuVersion.V2Classic => "The original classic",
            MikuVersion.NT => "The modern NT engine",
            MikuVersion.Sekai => "Project SEKAI version",
            _ => "Unknown version"
        };

        Console.WriteLine(description);

        // Pattern matching with is
        if (miku is { Energy: > 50, Mode: StageMode.Idle })
        {
            miku.Activate();
        }

        // Deconstruction
        var (name, version) = miku;
        Console.WriteLine($"Name: {name}, Version: {version}");

        // Try-catch with when clause
        try
        {
            await miku.PerformConcertAsync(songs);
        }
        catch (PerformanceException ex) when (ex.Message.Contains("energy"))
        {
            Console.WriteLine($"Energy issue: {ex.Message}");
        }

        // Null-conditional and null-coalescing
        var voiceName = miku.VoiceBank?.Name ?? "Unknown";

        // Span and ranges
        ReadOnlySpan<char> colorSpan = DigitalDiva.CanonicalColor.AsSpan();
        var hexPart = DigitalDiva.CanonicalColor[1..];

        // Raw string literals (C# 11)
        var json = """
            {
                "name": "Hatsune Miku",
                "version": "V2",
                "color": "#39C5BB"
            }
            """;

        Console.WriteLine(json);
        Console.WriteLine($"Performance complete! Energy: {miku.Energy}");
    }
}

File: examples/showcase.java
/**
 * Hatsune Miku Theme - Java Showcase
 * All-Miku Synthesis: Every voice, one stage.
 */

package com.mikutheme.showcase;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;

// Interface: #B2EBE7 Bold (Ice Teal - Append Light)
interface Synthesizer {
    // Method signature: #00BCD4 (NT Cyan)
    String sing(String song) throws PerformanceException;
    MikuVersion getVersion();
}

// Functional interface
@FunctionalInterface
interface SongProcessor {
    String process(String song);
}

// Enum: #B2EBE7 Bold, EnumMember: #E05096
enum MikuVersion {
    V2_CLASSIC("V2", 2007),
    V3("V3", 2013),
    V4X("V4X", 2016),
    NT("NT", 2020),
    SEKAI("SEKAI", 2020);

    // Fields: #90B8B2
    private final String code;
    private final int releaseYear;

    MikuVersion(String code, int releaseYear) {
        this.code = code;
        this.releaseYear = releaseYear;
    }

    public String getCode() {
        return code;
    }

    public int getReleaseYear() {
        return releaseYear;
    }
}

// Custom exception: #FF5370
class PerformanceException extends Exception {
    public PerformanceException(String message) {
        super(message);
    }
}

// Record (Java 16+): #B2EBE7 Bold
record VoiceBank(String name, MikuVersion version, int[] frequencyRange) {
    // Compact constructor
    public VoiceBank {
        Objects.requireNonNull(name, "Name cannot be null");
        if (frequencyRange.length != 2) {
            throw new IllegalArgumentException("Frequency range must have 2 elements");
        }
    }
}

// Sealed class (Java 17+)
sealed abstract class StageElement permits Performer, Prop, Light {
    abstract void activate();
}

final class Performer extends StageElement {
    @Override
    void activate() {
        System.out.println("Performer activated");
    }
}

final class Prop extends StageElement {
    @Override
    void activate() {
        System.out.println("Prop activated");
    }
}

final class Light extends StageElement {
    @Override
    void activate() {
        System.out.println("Light activated");
    }
}

// Generic class: #B2EBE7 Bold (Ice Teal)
public class DigitalDiva<T> implements Synthesizer {
    // Constants: #E05096 (Magenta LED - Rhythm)
    public static final String CANONICAL_COLOR = "#39C5BB";
    private static final int DEFAULT_BPM = 39;

    // Instance fields: #90B8B2
    private final String name;
    private final VoiceBank voiceBank;
    private int energy;
    private Map<String, T> metadata;

    // Constructor
    public DigitalDiva(String name, MikuVersion version) {
        this.name = name;
        this.voiceBank = new VoiceBank(name, version, new int[]{80, 1100});
        this.energy = 100;
        this.metadata = new HashMap<>();
    }

    // Method with annotation: #FFD740 Italic (Decorator)
    @Override
    public String sing(String song) throws PerformanceException {
        // Keywords: #39C5BB Bold
        if (energy < 10) {
            throw new PerformanceException("Low energy: Please recharge with leeks");
        }

        // String formatting: #9CCC65 (Negi Green)
        String message = String.format("[MIKU] Now singing: %s", song);
        System.out.println(message);

        energy -= 10;
        return message;
    }

    @Override
    public MikuVersion getVersion() {
        return voiceBank.version();
    }

    // Generic method: #00BCD4
    public <U extends Comparable<U>> U findMax(List<U> items) {
        return items.stream()
                .max(Comparator.naturalOrder())
                .orElseThrow();
    }

    // Varargs
    public void performMultiple(String... songs) throws PerformanceException {
        for (String song : songs) {
            sing(song);
        }
    }

    // Static method: #00BCD4 Underline
    public static String getCanonicalColor() {
        return CANONICAL_COLOR;
    }

    // Pattern matching (Java 21+)
    public String describeElement(StageElement element) {
        return switch (element) {
            case Performer p -> "Performer on stage";
            case Prop p -> "Stage prop";
            case Light l -> "Stage lighting";
        };
    }
}

// Main class for demonstration
class MikuShowcase {
    public static void main(String[] args) {
        // Local variable type inference (var)
        var miku = new DigitalDiva<String>("Hatsune Miku", MikuVersion.V2_CLASSIC);

        // Try-with-resources
        try (var executor = Executors.newVirtualThreadPerTaskExecutor()) {
            // Lambda expression: #00BCD4
            SongProcessor processor = song -> song.toUpperCase();

            // Method reference: #00BCD4
            List<String> songs = List.of("Melt", "World is Mine", "Rolling Girl");
            songs.forEach(System.out::println);

            // Stream API
            List<String> processedSongs = songs.stream()
                    .filter(s -> s.length() > 4)
                    .map(String::toUpperCase)
                    .sorted()
                    .collect(Collectors.toList());

            // CompletableFuture
            CompletableFuture<String> future = CompletableFuture
                    .supplyAsync(() -> "World is Mine")
                    .thenApply(processor::process)
                    .exceptionally(ex -> "Error: " + ex.getMessage());

            // Optional
            Optional<String> firstSong = songs.stream().findFirst();
            firstSong.ifPresent(song -> System.out.println("First: " + song));

        } catch (Exception e) {
            // Error handling: #FF5370
            System.err.println("Error: " + e.getMessage());
        }

        // Text block (Java 15+)
        String json = """
                {
                    "name": "Hatsune Miku",
                    "version": "V2",
                    "color": "#39C5BB"
                }
                """;

        System.out.println(json);
    }
}
